kubelet: Add GracefulNodeShutdownPodPolicy  config option to graceful node shutdown

bobbypage opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Graceful node shutdown currently places pods into terminal phase upon shutdown. As discussed in https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-982763592 and https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-982766908 some users/distributions would benefit from the ability to toggle this behavior and instead to not put pods into terminal phase on shutdown. 

For example, if it's expected that after shutdown the node will reboot, it may be desirable to not place the pods into terminal phase, so that way after reboot the pods will start running again after the node comes back up.

This PR adds a new kubelet configuration option to toggle this behavior `GracefulNodeShutdownPodPolicy`. The default setting of `GracefulNodeShutdownPodPolicy` is enabled `SetTerminal` (which matches current behavior since 1.20) when graceful node shutdown was introduced. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #108991

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add a new kubelet configuration option `GracefulNodeShutdownPodPolicy` to toggle setting pods to terminal phase during graceful node shutdown.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bobbypage committed:
kubelet: Use the GracefulNodeShutdownPodPolicy config

Pass the config option from the kubelet config to the graceful node
shutdown manager, and only update phase if the option is enabled.

Signed-off-by: David Porter <david@porter.me>

bobbypage committed:
test: Update node shutdown test for GracefulNodeShutdownPodPolicy

Refactor the existing test to ensure we can test two scenarios, of
`SetTerminal` or `LeaveRunning` which are two possible
GracefulNodeShutdownPodPolicy policy types.

Modify the check `isPodShutdown` to pass `expectPhaseToBeFailed` and
only check phase if the option is enabled.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
/test pull-kubernetes-node-kubelet-serial-containerd

bobbypage commented:
/test pull-kubernetes-node-kubelet-serial-containerd

yangjunmyfm192085 commented on test/e2e_node/node_shutdown_linux_test.go:
```diff
@@ -417,6 +365,108 @@ var _ = SIGDescribe("GracefulNodeShutdown [Serial] [NodeFeature:GracefulNodeShut
+	}, podStatusUpdateTimeout, pollInterval).Should(gomega.BeNil())
+
+	ginkgo.By("Verifying that all pods are shutdown")
+	// All pod should be shutdown
```
```suggestion
	// All pods should be shutdown
```

bobbypage replied:
Fixed

fedebongio commented:
/remove-sig api-machinery

bobbypage commented:
/cc @smarterclayton as per discussion in https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-982766908

bobbypage commented:
/cc @mrunalp

bobbypage commented:
/retest

mrunalp commented:
This looks fine to me. Thanks!

rphillips commented:
@bobbypage what is your thought of an override via the dbus shutdown command if the kubelet should set the pods to terminal or not? This would allow controllers managing the shutdown to decide if the Kubelet should set pods to terminal to false. (basically make this option dynamic).

ehashman commented:
/test pull-kubernetes-node-kubelet-serial-containerd

ehashman commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -407,6 +407,12 @@ type KubeletConfiguration struct {
 	// list when the node is shutting down.
 	ShutdownGracePeriodByPodPriority []ShutdownGracePeriodByPodPriority
+	// ShutdownSetPodsPhaseToTerminal sets pods to terminal phase during shutdown (enabled by default).
+	// If a node becomes ready again after graceful node shutdown (e.g. a reboot), then having ShutdownSetPodsPhaseToTerminal enabled, will result in pods not being started on the same node since the pods will be placed into terminal phase. In contrast, if ShutdownSetPodsPhaseToTerminal is disabled, pods will not be placed into terminal phase, and as result they will resume to run after the node bec ...
```
Please reformat this line to not exceed ~80 chars.

bobbypage replied:
Done.

ehashman commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -724,6 +724,12 @@ type KubeletConfiguration struct {
 	// +optional
 	ShutdownGracePeriodByPodPriority []ShutdownGracePeriodByPodPriority `json:"shutdownGracePeriodByPodPriority,omitempty"`
+	// ShutdownSetPodsPhaseToTerminal sets pods to terminal phase during shutdown (enabled by default).
+	// If a node becomes ready again after graceful node shutdown (e.g. a reboot), then having ShutdownSetPodsPhaseToTerminal enabled, will result in pods not being started on the same node since the pods will be placed into terminal phase. In contrast, if ShutdownSetPodsPhaseToTerminal is disabled, pods will not be placed into terminal phase, and as result they will resume to run after the node bec ...
```
See comment above - please reflow the line.

bobbypage replied:
Done.

ehashman commented on test/e2e_node/node_shutdown_linux_test.go:
```diff
@@ -417,6 +365,108 @@ var _ = SIGDescribe("GracefulNodeShutdown [Serial] [NodeFeature:GracefulNodeShut
 	})
 })
 
+func runTestShutdownPodsWithVariousGracePeriods(f *framework.Framework, expectPodPhaseToBeFailed bool) {
```
`Diff -w` from moved snippet above:

```diff
diff --git a/tmp/foo5 b/tmp/foo6
index b692047..0150bf8 100644
--- a/tmp/foo5
+++ b/tmp/foo6
@@ -1,3 +1,11 @@
+func runTestShutdownPodsWithVariousGracePeriods(f *framework.Framework, expectPodPhaseToBeFailed bool) {
+       const (
+               pollInterval                        = 1 * time.Second
+               podStatusUpdateTimeout              = 30 * time.Second
+               nodeStatusUpdateTimeout             = 30 * time.Second
+               nodeShutdownGracePeriod             = 20 * time.Second
+               nodeShutdownGracePeriodCriticalPods = 10 * time.Second
+       )
 
        nodeName := getNodeName(f)
        nodeSelector := fields.Set{
@@ -50,12 +58,12 @@
                        }
 
                        if kubelettypes.IsCriticalPod(&pod) {
-                                               if isPodShutdown(&pod) {
+                               if isPodShutdown(&pod, expectPodPhaseToBeFailed) {
                                        framework.Logf("Expecting critical pod to be running, but it's not currently. Pod: %q, Pod Status %+v", pod.Name, pod.Status)
                                        return fmt.Errorf("critical pod should not be shutdown, phase: %s", pod.Status.Phase)
                                }
                        } else {
-                                               if !isPodShutdown(&pod) {
+                               if !isPodShutdown(&pod, expectPodPhaseToBeFailed) {
                                        framework.Logf("Expecting non-critical pod to be shutdown, but it's not currently. Pod: %q, Pod Status %+v", pod.Name, pod.Status)
                                        return fmt.Errorf("pod should be shutdown, phase: %s", pod.Status.Phase)
                                }
@@ -65,7 +73,7 @@
        }, podStatusUpdateTimeout, pollInterval).Should(gomega.BeNil())
 
        ginkgo.By("Verifying that all pods are shutdown")
-                       // All pod should be shutdown
+       // All pods should be shutdown
        gomega.Eventually(func() error {
                list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{
                        FieldSelector: nodeSelector,
@@ -80,7 +88,7 @@
                                framework.Logf("Detected invalid pod state for pod %q: pod status: %+v", pod.Name, pod.Status)
                                framework.Failf("failing test due to detecting invalid pod status")
                        }
-                                       if !isPodShutdown(&pod) {
+                       if !isPodShutdown(&pod, expectPodPhaseToBeFailed) {
                                framework.Logf("Expecting pod to be shutdown, but it's not currently: Pod: %q, Pod Status %+v", pod.Name, pod.Status)
                                return fmt.Errorf("pod should be shutdown, phase: %s", pod.Status.Phase)
                        }
@@ -90,3 +98,4 @@
                // Critical pod starts shutdown after (nodeShutdownGracePeriod-nodeShutdownGracePeriodCriticalPods)
                podStatusUpdateTimeout+(nodeShutdownGracePeriod-nodeShutdownGracePeriodCriticalPods),
                pollInterval).Should(gomega.BeNil())
+}
```

:+1:

ehashman commented on test/e2e_node/node_shutdown_linux_test.go:
```diff
@@ -231,6 +140,45 @@ var _ = SIGDescribe("GracefulNodeShutdown [Serial] [NodeFeature:GracefulNodeShut
 		})
 	})
 
+	ginkgo.Context("when gracefully shutting down with ShutdownSetPodsPhaseToTerminal=false", func() {
```
Diff -w between test chunks:

```diff
diff --git a/tmp/foo8 b/tmp/foo7
index bf123ba..9d469fb 100644
--- a/tmp/foo8
+++ b/tmp/foo7
@@ -1,11 +1,11 @@
-ginkgo.Context("when gracefully shutting down with ShutdownSetPodsPhaseToTerminal=true", func() {
-
+       ginkgo.Context("when gracefully shutting down with ShutdownSetPodsPhaseToTerminal=false", func() {
                const (
                        pollInterval                        = 1 * time.Second
                        nodeStatusUpdateTimeout             = 30 * time.Second
                        nodeShutdownGracePeriod             = 20 * time.Second
                        nodeShutdownGracePeriodCriticalPods = 10 * time.Second
                )
+
                tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
                        initialConfig.FeatureGates = map[string]bool{
                                string(features.GracefulNodeShutdown):                   true,
@@ -13,7 +13,9 @@ ginkgo.Context("when gracefully shutting down with ShutdownSetPodsPhaseToTermina
                        }
                        initialConfig.ShutdownGracePeriod = metav1.Duration{Duration: nodeShutdownGracePeriod}
                        initialConfig.ShutdownGracePeriodCriticalPods = metav1.Duration{Duration: nodeShutdownGracePeriodCriticalPods}
+                       initialConfig.ShutdownSetPodsPhaseToTerminal = false
                })
+
                ginkgo.BeforeEach(func() {
                        if err := lookEmitSignalCommand(); err != nil {
                                e2eskipper.Skipf("skipping test because: %v", err)
@@ -22,6 +24,7 @@ ginkgo.Context("when gracefully shutting down with ShutdownSetPodsPhaseToTermina
                        ginkgo.By("Wait for the node to be ready")
                        waitForNodeReady()
                })
+
                ginkgo.AfterEach(func() {
                        ginkgo.By("Emitting Shutdown false signal; cancelling the shutdown")
                        err := emitSignalPrepareForShutdown(false)
@@ -29,5 +32,7 @@ ginkgo.Context("when gracefully shutting down with ShutdownSetPodsPhaseToTermina
                })
 
                ginkgo.It("should be able to gracefully shutdown pods with various grace periods", func() {
-                       runTestShutdownPodsWithVariousGracePeriods(f, true)
+                       runTestShutdownPodsWithVariousGracePeriods(f, false)
+               })
+
        })
```

:+1:

ehashman reviewed (commented):
LGTM for API, implementation, and test; one requested fix for the API docs.

ehashman commented:
/priority important-soon
/triage accepted

bobbypage commented:
> @bobbypage what is your thought of an override via the dbus shutdown command if the kubelet should set the pods to terminal or not? This would allow controllers managing the shutdown to decide if the Kubelet should set pods to terminal to false. (basically make this option dynamic).

hmm... interesting idea, but I'm not sure if it's worth the complications...I'm not sure of any other examples of kubelet config changing dynamically. This sounds like a workaround of dynamic kubelet configuration, which was deprecated because it was hard to reason about dynamically changing kubelet's config in first place :) Is there some specific use cases you have in mind where it would be useful to have this ability?

ehashman reviewed (commented):
/lgtm
/approve

ehashman commented:
/label api-review
/test pull-kubernetes-node-kubelet-serial-containerd

rphillips commented:
@bobbypage I was thinking it wouldn't need to be hard coded when the kubelet starts... not a blocker though on this PR

bobbypage commented:
/assign @liggitt 

for API review

bobbypage commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -407,6 +407,16 @@ type KubeletConfiguration struct {
+	// Default: true
+	// +featureGate=GracefulNodeShutdown
+	// +optional
+	ShutdownSetPodsPhaseToTerminal bool
```
Thanks @liggitt, makes sense to make this more extensible in case we want to add more options in the future.

> Does the current behavior just set terminal, or does it also set a deletionTimestamp on the pod? If the former, is setting a deletionTimestamp another possibility that should be represented?

The former only sets phase to terminal, ti does not set deletion timestamp. But I agree, it makes sense make this a string with policy types incase we want to add something down the road.

ehashman replied:
The thought had also occurred to me but I wasn't sure if it'd be blocking. Agreed!

bobbypage committed:
kubelet: add new config option GracefulNodeShutdownPodPolicy

Add a new kubelet config option, GracefulNodeShutdownPodPolicy to
control policy of how pod termination is handled during graceful
shutdown.

The two options are `SetTerminal` or `LeaveRunning`. The default is
`SetTerminal` whill place set pod phase to terminal when the node
gracefully shuts down (which is the current default behavior).

The other policy type, `LeaveRunning`, will not change the pod phase during
graceful shutdown.

Also run `make generated_files` to regen api files after kubelet
configuration update.

Signed-off-by: David Porter <david@porter.me>

bobbypage committed:
Add validation for GracefulNodeShutdownPodPolicy

Add validation which checks that GracefulNodeShutdownPodPolicy is set to
a expected value which is currently only `SetTerminal` or
`LeaveRunning`.

Also update the kubelet config validation unit tests.

Signed-off-by: David Porter <david@porter.me>

bobbypage commented:
Updated based on @liggitt recommendation of using a string of a boolean for  `GracefulNodeShutdownPodPolicy`, xref https://github.com/kubernetes/kubernetes/pull/108941#discussion_r836551199

ehashman commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -407,6 +421,17 @@ type KubeletConfiguration struct {
+	// since the pods will be placed into terminal phase.
+	// In contrast, with `LeaveRunning` policy, pods will not be placed into terminal phase,
+	// and as result they will resume to run after the node becomes ready again.
+	// Default: true
```
Please update the default here

```suggestion
	// Default: SetTerminal
```

bobbypage replied:
Thanks, updated.

ehashman commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -724,6 +738,17 @@ type KubeletConfiguration struct {
+	// since the pods will be placed into terminal phase.
+	// In contrast, with `LeaveRunning` policy, pods will not be placed into terminal phase,
+	// and as result they will resume to run after the node becomes ready again.
+	// Default: true
```
Same comment here re: default

bobbypage replied:
Updated

bobbypage commented:
/retest

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -75,6 +75,20 @@ const (
+
+	// GracefulNodeShutdownPodPolicyLeaveRunning is Graceful Node Shutdown policy type,
+	// which will ensure that during graceful node shutdown,
+	// pods will be left running and will start again if the if nodes becomes ready again.
```
should the docs distinguish between what is happening to the process and what is marked in the API?

```suggestion
	// pods will be left marked as running and will start again if the node becomes ready again.
```

bobbypage replied:
Good suggestion, I rephrased to emphaize that kubelet will terminate the pods in both cases, and the status to API server is different depending between `SetTerminal` `LeaveRunning`

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -75,6 +75,20 @@ const (
+const (
+	// GracefulNodeShutdownPodPolicySetTerminal is Graceful Node Shutdown policy type,
+	// which will ensure that during graceful node shutdown,
+	// pods will be placed into terminal phase and will start again if nodes becomes ready again.
```
also, if the behavior this is driving is to set pods that have not already succeeded to `status.phase="Failed"`, should this be called `...SetFailed`? be as specific as possible in the docs, even down to describing the API field in the pod that is set, and under what conditions ("pods that are not in Succeeded or Failed phase will be set to Failed", etc)

liggitt replied:
propagate any doc changes to v1beta1/types.go as well

bobbypage replied:
oops, this is missing a `not`, i.e. `s/will start/will not start/`. Fixed.

The current behavior is to set it to `Failed` (I added note of this in the doc comments), but I think longer term we want to improve this and set to `Succeeded` if the pod succeeded in existing, so I think it would better to keep it more general (i.e. `SetTerminal`) which refers any type of terminal phase.

liggitt replied:
> longer term we want to improve this and set to Succeeded if the pod succeeded in existing

I'm trying to decide if that change in behavior could break anyone, and if we'd want to make that a distinct behavior

liggitt replied:
I'll defer to @smarterclayton on the compatibility question around setting Terminal / Failed / Succeeded

liggitt commented:
structure lgtm, a couple questions on the docs and the name/value for the SetTerminal option

bobbypage commented:
Updated doc comments in latest push.

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -75,6 +75,22 @@ const (
+	// GracefulNodeShutdownPodPolicySetTerminal is Graceful Node Shutdown policy type,
+	// which will ensure that during graceful node shutdown, pods will be terminated by kubelet and
+	// placed into terminal Failed phase on API server,
+	// and as a result, will not start again if nodes becomes ready again.
```
nit: clarify its a particular node we're talking about becoming ready again

```suggestion
	// and as a result, will not start again if the node becomes ready again.
```

bobbypage replied:
Fixed, thanks.

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -75,6 +75,22 @@ const (
+	// GracefulNodeShutdownPodPolicyLeaveRunning is Graceful Node Shutdown policy type,
+	// which will ensure that during graceful node shutdown,
+	// pods will be terminated by kubelet, but left running on API server,
+	// and a result, will start again if the if nodes becomes ready again.
```
```suggestion
	// and a result, will start again if the node becomes ready again.
```

bobbypage replied:
Fixed, thanks.

smarterclayton commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -407,6 +423,17 @@ type KubeletConfiguration struct {
+	// If a node becomes ready again after graceful node shutdown (e.g. a reboot), will result in pods not being started on the same node
+	// since the pods will be placed into terminal phase. Pods will be placed in the Failed terminal phase on the API server.
+	// In contrast, with `LeaveRunning` policy, pods will not be placed into terminal phase,
+	// and as result they will resume to run after the node becomes ready again.
```
I would probably say this should document that the default behavior for Kubernetes is `LeaveRunning`.  Also, we should describe how the Kubelet detects that the node is being shut down, since the sequence here was:

1. shutdown of node didn't do anything
2. we added a feature that behaved like SetTerminal (different than 1)
3. we realized we needed 1, so we added this flag
4. users seeing this flag don't know how to go from 1 (the way it worked from 1.0 to 1.24) to 3 correctly, or they might be confused that 2 is how it works.

liggitt replied:
> I would probably say this should document that the default behavior for Kubernetes is LeaveRunning

That's a good point. Requiring opt in for the GracefulNodeShutdown feature to mark pods as terminal makes sense to me.

> Also note - I don't expect this flag to have a default value per se - I expect the presence of this flag OR shutdown grace period to turn the feature on (makes us deviate from behavior 1), and thus this value is only defaulted to SetTerminal when its omitted AND ShutdownGracePeriod is set.

That's another good point. I think that means we wouldn't use official API defaulting, but would instead accept "" and treat it like `LeaveRunning` when initializing the kubelet runtime options

bobbypage replied:
Thanks @smarterclayton and @liggitt for the feedback.

> Requiring opt in for the GracefulNodeShutdown feature to mark pods as terminal makes sense to me. 

This is already the case, `GracefulNodeShutdownPodPolicy` will only be used if graceful node is enabled (i.e. `ShutdownGracePeriod` is set). Otherwise, if graceful node shutdown is not enabled, shutdown of node won't do anything.

Does it make sense as follow up to then:

1) Only default `GracefulNodeShutdownPodPolicy` to `SetTerminal` if Graceful Node Shutdown feature is enabled (i.e. `ShutdownGracePeriod`) in `pkg/kubelet/apis/config/v1beta1/defaults.go`. Otherwise the setting will be left unset, `""`. 
2) Update the API docs to make it clear that `GracefulNodeShutdownPodPolicy` will be defaulted to `SetTerminal` only if Graceful Node shutdown is enabled.

bobbypage replied:
Updated as mentioned above. Please take another look, thanks!

smarterclayton replied:
> Only default GracefulNodeShutdownPodPolicy to SetTerminal if Graceful Node Shutdown feature is enabled (i.e. ShutdownGracePeriod) in pkg/kubelet/apis/config/v1beta1/defaults.go. Otherwise the setting will be left unset, "".

Let me translate this to make sure we're all on the same page with possible configurations

1. User specifies `ShutdownGracePeriod` but `GracefulNodeShutdownPodPolicy` is empty or null -> defaults to `SetTerminal` (preserves behavior for beta users)
2. User specifies `GracefulNodeShutdownPodPolicy` but `ShutdownGracePeriod` is empty or zero -> default to `LeaveRunning`
3. User specifies `GracefulNodeShutdownPodPolicy` and `ShutdownGracePeriod` -> no default necessary

If we did this, then I think users who move from 2 to 3 would be confused (they'd switch from LeaveRunning to SetTerminal, which is bad).

So I might argue that 2 should be:

2. User specifies `GracefulNodeShutdownPodPolicy` but `ShutdownGracePeriod` is empty or zero -> validation error, user must explicitly specify `ShutdownGracePeriod`

Which then trains new users to make a choice (and we can say `LeaveRunning is the default behavior for pods on kubelet shutdown when grace period is disabled` or something).

smarterclayton commented:
Some api doc comments

pacoxu commented:
/cc @wzshiming

endocrimes commented:
Are we still targeting this for 1.25?
